File: dqops/src/main/java/com/dqops/rest/models/metadata/DefaultColumnChecksPatternListModel.java
/*
 * Copyright © 2021 DQOps (support@dqops.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dqops.rest.models.metadata;

import com.dqops.metadata.defaultchecks.column.ColumnDefaultChecksPatternSpec;
import com.dqops.metadata.defaultchecks.column.TargetColumnPatternSpec;
import com.dqops.utils.docs.generators.SampleStringsRegistry;
import com.dqops.utils.docs.generators.SampleValueFactory;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import io.swagger.annotations.ApiModel;
import lombok.Data;

/**
 * The listing model of column-level default check patterns that is returned by the REST API.
 */
@Data
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@ApiModel(value = "DefaultColumnChecksPatternListModel", description = "Default column-level checks pattern list model")
public class DefaultColumnChecksPatternListModel {
    /**
     * Pattern name.
     */
    @JsonPropertyDescription("Pattern name.")
    private String patternName;

    @JsonPropertyDescription("The priority of the pattern. Patterns with lower values are applied before patterns with higher priority values.")
    private int priority;

    /**
     * Target column filters.
     */
    @JsonPropertyDescription("The filters for the target column.")
    private TargetColumnPatternSpec targetColumn;

    /**
     * Boolean flag that decides if the current user can update or delete this object.
     */
    @JsonPropertyDescription("Boolean flag that decides if the current user can update or delete this object.")
    private boolean canEdit;

    /**
     * Optional parsing error that was captured when parsing the YAML file.
     * This field is null when the YAML file is valid. If an error was captured, this field returns the file parsing error message and the file location.
     */
    @JsonPropertyDescription("Optional parsing error that was captured when parsing the YAML file. " +
            "This field is null when the YAML file is valid. If an error was captured, this field returns the file parsing error message and the file location.")
    private String yamlParsingError;

    public DefaultColumnChecksPatternListModel() {
    }

    /**
     * Creates a default table-level checks pattern model from a specification by cherry-picking relevant fields.
     * @param checksPatternSpec Source default checks pattern specification.
     * @param isEditor       The current user has the editor permission.
     * @return Default checks pattern list model.
     */
    public static DefaultColumnChecksPatternListModel fromPatternSpecification(
            ColumnDefaultChecksPatternSpec checksPatternSpec,
            boolean isEditor) {
        return new DefaultColumnChecksPatternListModel() {{
            setPatternName(checksPatternSpec.getPatternName());
            setPriority(checksPatternSpec.getPriority());
            setTargetColumn(checksPatternSpec.getTarget());
            setCanEdit(isEditor);
            setYamlParsingError(checksPatternSpec.getYamlParsingError());
        }};
    }

    public static class ColumnDefaultChecksPatternListModelSampleFactory implements SampleValueFactory<DefaultColumnChecksPatternListModel> {
        @Override
        public DefaultColumnChecksPatternListModel createSample() {
            return new DefaultColumnChecksPatternListModel() {{
                setPatternName(SampleStringsRegistry.getPatternName());
                setPriority(100);
                setTargetColumn(new TargetColumnPatternSpec() {{
                    setConnection("dwh");
                    setColumn("id");
                }});
                setCanEdit(true);
            }};
        }
    }
}

File: distribution/python/dqops/client/models/default_column_checks_pattern_list_model.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.target_column_pattern_spec import TargetColumnPatternSpec


T = TypeVar("T", bound="DefaultColumnChecksPatternListModel")


@_attrs_define
class DefaultColumnChecksPatternListModel:
    """Default column-level checks pattern list model

    Attributes:
        pattern_name (Union[Unset, str]): Pattern name.
        priority (Union[Unset, int]): The priority of the pattern. Patterns with lower values are applied before
            patterns with higher priority values.
        target_column (Union[Unset, TargetColumnPatternSpec]):
        can_edit (Union[Unset, bool]): Boolean flag that decides if the current user can update or delete this object.
        yaml_parsing_error (Union[Unset, str]): Optional parsing error that was captured when parsing the YAML file.
            This field is null when the YAML file is valid. If an error was captured, this field returns the file parsing
            error message and the file location.
    """

    pattern_name: Union[Unset, str] = UNSET
    priority: Union[Unset, int] = UNSET
    target_column: Union[Unset, "TargetColumnPatternSpec"] = UNSET
    can_edit: Union[Unset, bool] = UNSET
    yaml_parsing_error: Union[Unset, str] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        pattern_name = self.pattern_name
        priority = self.priority
        target_column: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.target_column, Unset):
            target_column = self.target_column.to_dict()

        can_edit = self.can_edit
        yaml_parsing_error = self.yaml_parsing_error

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if pattern_name is not UNSET:
            field_dict["pattern_name"] = pattern_name
        if priority is not UNSET:
            field_dict["priority"] = priority
        if target_column is not UNSET:
            field_dict["target_column"] = target_column
        if can_edit is not UNSET:
            field_dict["can_edit"] = can_edit
        if yaml_parsing_error is not UNSET:
            field_dict["yaml_parsing_error"] = yaml_parsing_error

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.target_column_pattern_spec import TargetColumnPatternSpec

        d = src_dict.copy()
        pattern_name = d.pop("pattern_name", UNSET)

        priority = d.pop("priority", UNSET)

        _target_column = d.pop("target_column", UNSET)
        target_column: Union[Unset, TargetColumnPatternSpec]
        if isinstance(_target_column, Unset):
            target_column = UNSET
        else:
            target_column = TargetColumnPatternSpec.from_dict(_target_column)

        can_edit = d.pop("can_edit", UNSET)

        yaml_parsing_error = d.pop("yaml_parsing_error", UNSET)

        default_column_checks_pattern_list_model = cls(
            pattern_name=pattern_name,
            priority=priority,
            target_column=target_column,
            can_edit=can_edit,
            yaml_parsing_error=yaml_parsing_error,
        )

        default_column_checks_pattern_list_model.additional_properties = d
        return default_column_checks_pattern_list_model

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
